kyaml/yaml/walk/associative_sequence.go

```diff
@@ -310,66 +310,136 @@ func (l Walker) elementKey() (string, error) {
+				returnValues = append(returnValues[:p+1], append([][]string{value}, returnValues[p+1:]...)...)
+			}
+
+			// Keep track of the previous value within this source, so that subsequent values can be inserted in
+			// the correct position
+			previous = value
 		}
 	}
+
 	return returnValues
 }
 
 // elementPrimitiveValues returns the primitive values in an associative list -- eg. finalizers
 func (l Walker) elementPrimitiveValues() [][]string {
-	// use slice to to keep elements in the original order
-	var returnValues [][]string
-	seen := sets.String{}
-	// if we are doing append, dest node should be the first.
-	// otherwise dest node should be the last.
-	beginIdx := 0
-	if l.MergeOptions.ListIncreaseDirection == yaml.MergeOptionsListPrepend {
-		beginIdx = 1
+	returnValues := make([][]string, 0)
+
+	// Start by adding the elements from the destination (index 0)
+	src := l.Sources[0]
+	items := src.YNode().Content
+	for _, item := range items {
+		value := []string{item.Value}
+		returnValues = append(returnValues, value)
 	}
-	for i := range l.Sources {
-		src := l.Sources[(i+beginIdx)%len(l.Sources)]
-		if src == nil {
-			continue
-		}
 
-		// add the value of the field for each element
-		// don't check error, we know this is a list node
-		for _, item := range src.YNode().Content {
-			if seen.Has(item.Value) {
-				continue
+	// Iterate through each of the remaining sources and update returnValues by appending, prepending, and reordering
+	patchSources := l.Sources[1:]
+	for _, src := range patchSources {
+		items = src.YNode().Content
+		var previous []string
+		for _, item := range items {
+			value := []string{item.Value}
+			// Attempt to locate this value in returnValues
+			i := indexOf(returnValues, value)
+
+			if previous == nil {
+				// If the value dost not yet exist and there is no previous value, then add it to the beginning or end
+				// of returnValues, depending on ListIncreaseDirection
+				if i == -1 {
+					if l.MergeOptions.ListIncreaseDirection == yaml.MergeOptionsListPrepend {
+						returnValues = append([][]string{value}, returnValues...)
+					} else {
+						returnValues = append(returnValues, value)
+					}
+				}
+			} else {
+				// If there is a previous value and this value already exists in returnValues, remove it and reinsert
+				// it at the position directly after the previous value. If the value does not yet exist in
+				// returnValues, insert it directly after the previous value.
+				if i >= 0 {
+					returnValues = append(returnValues[:i], returnValues[i+1:]...)
+				}
+				p := indexOf(returnValues, previous)
+				returnValues = append(returnValues[:p+1], append([][]string{value}, returnValues[p+1:]...)...)
 			}
-			returnValues = append(returnValues, []string{item.Value})
-			seen.Insert(item.Value)
+
+			// Keep track of the previous value within this source, so that subsequent values can be inserted in
+			// the correct position
+			previous = value
```

Comment:
Can you factor L335-351 and L388-409 to a function so it's not copy/pasted?

To avoid the situation down the road where one copy is changed but  not the other.